### ISSUE TITLE
[2.x] Migration plan of key plugins

### ISSUE THREAD

**Issue by adpi2** [sbt/sbt#7698]
I create this issue to keep track of the knowledge about migration of plugins to sbt 2.x

**Comment by adpi2** [comment]
[Here](https://github.com/sbt/sbt-buildinfo/compare/develop...adpi2:sbt-buildinfo:sbt-2.x) I migrated [sbt/sbt-buildinfo](https://github.com/sbt/sbt-buildinfo) to 2.0.0-alpha12-SNAPSHOT.

The main difficulty was the replacement of `Manifest` by `Class` caused by https://github.com/sbt/sbt/pull/7490, but it is scoped to a single file.

**Comment by adpi2** [comment]
[Here](https://github.com/scala-js/scala-js/compare/main...adpi2:scala-js:sbt-2.x) I migrated [scala-js/scala-js](https://github.com/scala-js/scala-js) to 2.0.0-alpha12-SNAPSHOT.

Some encountered difficulties are:
- Migration of the Scala.js linker to Scala 3, the build depends on it indirectly. Temporarily and for some dependencies, we use `CrossVersion.for3Use2_13`.
- Migration to slash syntax. sbt-scalajs still uses the old syntax to maintain compatibility with old sbt 1.x versions.
- Conversion from `HashedVirtulaFileRef` to `File` and vice-versa. There are about 26 places where I used the `fileConverter`. This alone probably makes the cross-compilation of `sbt-scalajs` almost impossible. The related keys are `apiMappings`, `packageBin`, `artifactPath`, and all the various `*Classpath`.
- Serialization and de-serialization of extra metadata in `Attributed`, about 10 occurrences. It was an `AttributeMap` and it's a `StringAttributeMap` now.
- `ClasspathDependency.apply` returns a `ClasspathDep[ProjectReference]`
- Problem of inference [here](https://github.com/adpi2/scala-js/blob/eb2d156ffd6c2562f2c13a9ceed84b0f9503b3fe/project/Build.scala#L1281-L1286). The compiler fails at inferring the conversion from `TaskKey[A]` to `BuildInfoKey.Entry[A]`.

I managed to load the build but it still fails to compile anything to Scala.js. I suspect there is a conflict in the classpath between the `scalajs-ir` brought by the Scala 3 compiler and the local one. Probably sbt should not have the Scala 3 compiler in its main classpath.

**Comment by jtjeferreira** [comment]
Hi @adpi2 ,

Do you have a list of the "key plugins"? I would like to help port a few of them. Is `2.0.0-alpha12-SNAPSHOT` published somewhere, or do I have to publish locally first?

**Comment by eed3si9n** [comment]
@adpi2 I started migration guide on sbt/website. Could you update your findings to the doc plz? - https://www.scala-sbt.org/2.x/docs/en/changes/migrating-from-sbt-1.x.html

@jtjeferreira Hopefully we'll have M1 sometime soon. This is more like a pre-migration check phase.

**Comment by adpi2** [comment]
Hi @jtjeferreira

Any plugin is a good plugin, but the most important for us are probably:
- The ones tested in the sbt CI itself: https://github.com/sbt/sbt/tree/develop/sbt-app/src/sbt-test/plugins
- The ones under the sbt organization: https://github.com/sbt?q=&type=all&language=&sort=stargazers
- The most popular ones: https://index.scala-lang.org/search?platforms=sbt1.0

**Comment by eed3si9n** [comment]
- https://github.com/sbt/sbt/wiki/sbt-2.x-plugin-migration

**Comment by kubukoz** [comment]
- Opened an issue in sbt-bloop https://github.com/scalacenter/bloop/issues/2455 and a PR in bloop-config to unblock: https://github.com/scalacenter/bloop/issues/2455
- sbt-tpolecat PR: https://github.com/typelevel/sbt-tpolecat/pull/239 (currently failing due to a compiler crash)

I think it'd be helpful to gather a list of dependencies between the plugins, so that we can more easily identify the "leaf" ones.

**Comment by kubukoz** [comment]
> In sbt 2.x, if you cross build an sbt plugin with Scala 3.x and 2.12.x, it will automatically cross build against sbt 1.x and sbt 2.x

based on this note from https://www.scala-sbt.org/2.x/docs/en/changes/migrating-from-sbt-1.x.html, I was hoping for this to be enough: `crossScalaVersions := Seq("2.12.19", "3.3.4")` - but it doesn't set `sbtVersion`. I'm doing this in a 1.10.2 build, is that expected? (let me know if I should make another issue, but this seems relevant to migrating plugins)

**Comment by eed3si9n** [comment]
Yea, that's why I added a comment in the code saying `// using sbt 2.x`.

Using sbt 1.x we need to implement the trick by hand like this:

```scala
lazy val scala212 = "2.12.20"
lazy val scala3 = "3.3.4"
ThisBuild / crossScalaVersions := Seq(scala212, scala3)

lazy val plugin = (project in file("sbt-pgp"))
  .enablePlugins(SbtPlugin)
  .settings(
    ....,
    (pluginCrossBuild / sbtVersion) := {
      scalaBinaryVersion.value match {
        case "2.12" => "1.5.8"
        case _      => "2.0.0-M2"
      }
    }
  )
```

I'll add that to the doc.

**Comment by kubukoz** [comment]
> Yea, that's why I added a comment in the code saying `// using sbt 2.x`.

yeah, it tracks that I'd miss something important like this 😅

**Comment by agboom** [comment]
Hi, thanks for doing all this work for sbt 2.x :smile:

I'd like to help in some way. Would it make plugin migrations easier to have the scala3-migrate plugin available for 2.x? 

I did a [first attempt to migrate the plugin](https://github.com/agboom/scala3-migrate/actions/runs/11201166740/job/31135764794#step:4:99). I managed to bootstrap the scala3-migrate plugin into the sbt plugin by (temporarily) disabling the `SbtPlugin` and adding the necessary sbt dependencies for it to compile. Then I ran `sbt migrateDependencies` and `sbt migrateSyntax` as instructed by the plugin. The compiler gives some classpath and macro related errors that I'm not familiar with.

Here's the full project build: https://github.com/agboom/scala3-migrate/blob/main/plugin/build.sbt

I do however realize that sbt is making the leap from Scala 2.12 to 3, and migrating is a lot easier (if only possible) from 2.13. So would this even help to migrate sbt plugins?

**Comment by agboom** [comment]
Update: I think the reason the macro compiler errors appear is because they don't work with TASTy reader unless the macros are defined for both Scala 2 and 3. At least that's what I read from [this article](https://www.scala-lang.org/blog/2020/11/19/scala-3-forward-compat.html#forward-compatibility).

For the classpath error, I tried to add `("org.scala-sbt" %% "librarymanagement-coursier"    % "2.0.0-alpha8").cross(CrossVersion.for2_13Use3)` as dependency (which I believes comes from [here](https://github.com/adpi2/sbt-coursier/tree/wip/sbt-2.x)), but that didn't make the error go away. Edit: looking at the dependency tree, `librarymanagement-coursier` is already added via sbt, so it doesn't make any difference anyway.

**Comment by mkurz** [comment]
For Play you can add the PR
- https://github.com/playframework/playframework/pull/12897

to the wiki page.
However Play depends on quite some plugins itself ~which need to be migrated first before we can do any real work really.~

Also, please add following plugins to the wiki pages, they are currently missing:
- [lightbend-labs/mima](https://github.com/lightbend-labs/mima) (457 stars)
- [sbt/sbt-ci-release](https://github.com/sbt/sbt-ci-release) (283 stars) (you mention it, but you don't have that in the list actually)
- [scalameta/sbt-scalafmt](https://github.com/scalameta/sbt-scalafmt) (196 stars)
- [johanandren/sbt-akka-version-check](https://github.com/johanandren/sbt-akka-version-check) (10 stars)
- [lightbend/sbt-bill-of-materials](https://github.com/lightbend/sbt-bill-of-materials) (5 stars)
- [Philippus/sbt-pekko-version-check](https://github.com/Philippus/sbt-pekko-version-check) (2 stars)

(Play all depends on them)

**Comment by mkurz** [comment]
For Twirl ~I opened~
- ~https://github.com/playframework/twirl/pull/841~ (replaced by https://github.com/playframework/twirl/pull/840)

Following plugins are missing on your wiki page:
- [scalacenter/bloop (sbt-bloop)](https://github.com/scalacenter/bloop) (902 stars)
- [portable-scala/sbt-crossproject](https://github.com/portable-scala/sbt-crossproject) (236 stars)
- [sbt/sbt-maven-plugin](https://github.com/sbt/sbt-maven-plugin) (0 stars)

**Comment by eed3si9n** [comment]
Feel free to edit the wiki page as you see fit.

The table was semi-programatically generated from the sbt documentation. Likely those plugins should be added to the plugin list as well.

**Comment by adpi2** [comment]
@agboom

> I'd like to help in some way. Would it make plugin migrations easier to have the scala3-migrate plugin available for 2.x?

In general, I don't think we can use `scala3-migrate` to migrate sbt plugins, because `scala3-migrate` assumes the initial scala version is 2.13, not 2.12, also because there are no common versions of sbt that is binary compatible with 2.13 and 3.x.

> I do however realize that sbt is making the leap from Scala 2.12 to 3, and migrating is a lot easier (if only possible) from 2.13. So would this even help to migrate sbt plugins?

Yes, but we don't have an 2.13 version of sbt, and even if we do, it would be too much work to migrate the ecosystem twice, from 2.12 to 2.13 and then to 3.x.

To help with the migration, you can find a plugin which is a leaf in https://github.com/sbt/sbt/wiki/sbt-2.x-plugin-migration, and try to migrate by hand following the instructions in https://www.scala-sbt.org/2.x/docs/en/changes/migrating-from-sbt-1.x.html#cross-building-sbt-plugin-with-sbt-1x

**Comment by agboom** [comment]
@adpi2 Thanks for your reply, that makes sense. I slowly got into that realisation when I was working on the migration. I guess migrating by hand isn't so bad, I'll try to find a leaf plugin.

**Comment by agboom** [comment]
Is there a way I can add [this PR in sbt-git](https://github.com/sbt/sbt-git/pull/255) to the [wiki](https://github.com/sbt/sbt/wiki/sbt-2.x-plugin-migration)?

**Comment by mkurz** [comment]
> Feel free to edit the wiki page as you see fit.

I don't have write access to the wiki (I do not ask for it, just saying)

**Comment by eed3si9n** [comment]
ok. Added you guys to https://github.com/orgs/sbt/teams/sbt-park-rangers/members team. Hopefully this allows you to edit the wiki.

**Comment by mkurz** [comment]
> ok. Added you guys to https://github.com/orgs/sbt/teams/sbt-park-rangers/members team. Hopefully this allows you to edit the wiki.

Still can't edit the wiki.

**Comment by eed3si9n** [comment]
The wiki was set up to create push rights, but I've unchecked the checkbox now, so now all GitHub users can edit it, I think.

**Comment by mkurz** [comment]
> The wiki was set up to create push rights, but I've unchecked the checkbox now, so now all GitHub users can edit it, I think.

Thanks it's working now, added Twirl which [is done](https://repo1.maven.org/maven2/org/playframework/twirl/sbt-twirl_sbt2.0.0-M2_3/2.1.0-M2/).

**Comment by eed3si9n** [comment]
I'm guessing that at this point general plugin migration issue can be closed? (see https://github.com/sbt/sbt/wiki/sbt-2.x-plugin-migration for individual progress)
If there are specific issue that should be fixed, we should open a new issue.
